File: run-brk.py
import json
import os
import sys
import subprocess
import random
import math
from pathlib import Path
from datetime import datetime
from zoneinfo import ZoneInfo

script_dir = os.path.dirname(os.path.abspath(__file__))
secrets_path = os.path.join(script_dir, 'secrets.json')

home = Path.home()
user_site = home / '.local' / 'lib' / f'python{sys.version_info.major}.{sys.version_info.minor}' / 'site-packages'
if user_site.exists() and str(user_site) not in sys.path:
    sys.path.insert(0, str(user_site))


def load_api_keys():
    keys = [
        "1dbc799fe411479686be512ee797a777",
        "8e13917d80264a63935d62aa2c448076",
    ]

    if os.path.exists(secrets_path):
        try:
            with open(secrets_path) as f:
                secrets = json.load(f)
            secret_key = secrets.get('TWELVEDATA_API_KEY')
            if secret_key:
                keys.insert(0, secret_key)
        except Exception:
            pass

    seen = set()
    deduped = []
    for key in keys:
        if key and key not in seen:
            deduped.append(key)
            seen.add(key)
    return deduped


API_KEYS = load_api_keys()


def interval_to_yfinance(interval):
    mapping = {
        '1min': '1m',
        '5min': '5m',
        '1h': '60m',
        '1day': '1d',
    }
    return mapping.get(interval, '1d')


def interval_to_minutes(interval):
    mapping = {
        '1min': 1,
        '5min': 5,
        '1h': 60,
        '1day': 390,
    }
    return mapping.get(interval, 5)


def yfinance_range_for_interval(interval, periods):
    try:
        periods = int(periods)
    except Exception:
        periods = 100

    if interval == '1min':
        return '7d' if periods <= 390 * 7 else '30d'
    if interval == '5min':
        return '60d'
    if interval == '1h':
        return '730d'
    return '5y'


def compute_output(ticker, interval, display, periods, values, source):
    current = float(values[0]['close'])
    highs = [float(d.get('high', 0)) for d in values if d.get('high') is not None]
    lows = [float(d.get('low', 0)) for d in values if d.get('low') is not None]

    highs_unique = sorted(set(highs), reverse=True)
    upper_resistances = [h for h in highs_unique if h > current][:2]

    lows_unique = sorted(set(lows))
    lower_resistances = [l for l in lows_unique if l < current][-2:][::-1]

    dist_to_upper = upper_resistances[0] - current if upper_resistances else float('inf')
    dist_to_lower = current - lower_resistances[0] if lower_resistances else float('inf')

    inv_upper = 1 / dist_to_upper if dist_to_upper > 0 else float('inf')
    inv_lower = 1 / dist_to_lower if dist_to_lower > 0 else float('inf')

    total_inv = inv_upper + inv_lower
    up_prob = round((inv_upper / total_inv) * 100, 1) if total_inv > 0 else 0
    down_prob = round((inv_lower / total_inv) * 100, 1) if total_inv > 0 else 0

    latest = values[0] if values else {}
    timestamp_raw = latest.get('datetime') or latest.get('timestamp')
    eastern_time = None
    eastern_label = 'US/Eastern'
    if timestamp_raw:
        try:
            if isinstance(timestamp_raw, (int, float)):
                dt_utc = datetime.fromtimestamp(timestamp_raw, tz=ZoneInfo('UTC'))
            else:
                parsed = str(timestamp_raw).replace('Z', '+00:00')
                dt = datetime.fromisoformat(parsed)
                dt_utc = dt if dt.tzinfo else dt.replace(tzinfo=ZoneInfo('UTC'))
            dt_et = dt_utc.astimezone(ZoneInfo('America/New_York'))
            eastern_time = dt_et.strftime('%Y-%m-%d %I:%M %p')
            eastern_label = dt_et.tzname() or 'US/Eastern'
        except Exception:
            eastern_time = str(timestamp_raw)

    volumes = [float(v.get('volume', 0) or 0) for v in values if v.get('volume') is not None]
    current_volume = float(latest.get('volume', 0) or 0)
    expected_volume = None
    volume_ratio = None
    if volumes:
        sample = volumes[1:] if len(volumes) > 1 else volumes
        if sample:
            expected_volume = sum(sample) / len(sample)
            if expected_volume > 0:
                volume_ratio = current_volume / expected_volume

    interval_minutes = interval_to_minutes(interval)
    expected_day_volume = None
    current_day_volume = None
    day_volume_ratio = None
    if interval != '1day' and volumes:
        bars_per_day = max(1, round(390 / interval_minutes))
        today_bars = volumes[:bars_per_day]
        prior_bars = volumes[bars_per_day:bars_per_day * 6]
        if today_bars:
            current_day_volume = sum(today_bars)
        if prior_bars:
            chunks = [prior_bars[i:i + bars_per_day] for i in range(0, len(prior_bars), bars_per_day)]
            full_chunks = [chunk for chunk in chunks if len(chunk) >= max(1, math.floor(bars_per_day * 0.6))]
            if full_chunks:
                day_totals = [sum(chunk) for chunk in full_chunks]
                expected_day_volume = sum(day_totals) / len(day_totals)
        if current_day_volume is not None and expected_day_volume and expected_day_volume > 0:
            day_volume_ratio = current_day_volume / expected_day_volume
    elif interval == '1day' and volumes:
        current_day_volume = current_volume
        sample = volumes[1:] if len(volumes) > 1 else volumes
        if sample:
            expected_day_volume = sum(sample) / len(sample)
            if expected_day_volume > 0:
                day_volume_ratio = current_day_volume / expected_day_volume

    return {
        'ticker': ticker,
        'display': display,
        'periods': int(periods),
        'interval': interval,
        'current_price': current,
        'focus_price': current,
        'quote_time_eastern': eastern_time,
        'quote_timezone': eastern_label,
        'upper_resistances': [{'price': r, 'diff': round(r - current, 2)} for r in upper_resistances],
        'lower_supports': [{'price': r, 'diff': round(current - r, 2)} for r in lower_resistances],
        'probabilities': {
            'up': up_prob,
            'down': down_prob
        },
        'volume': {
            'current_bar': round(current_volume, 2) if current_volume is not None else None,
            'expected_bar': round(expected_volume, 2) if expected_volume is not None else None,
            'bar_ratio': round(volume_ratio, 2) if volume_ratio is not None else None,
            'current_day': round(current_day_volume, 2) if current_day_volume is not None else None,
            'expected_day': round(expected_day_volume, 2) if expected_day_volume is not None else None,
            'day_ratio': round(day_volume_ratio, 2) if day_volume_ratio is not None else None,
        },
        'source': source,
    }


def fetch_twelvedata(ticker, interval, periods):
    api_keys = API_KEYS[:]
    random.shuffle(api_keys)

    last_error = 'No Twelve Data API keys configured'
    for apikey in api_keys:
        url = f'https://api.twelvedata.com/time_series?symbol={ticker}&interval={interval}&apikey={apikey}&outputsize={periods}'
        response = subprocess.run(['curl', '-s', url], capture_output=True, text=True)

        try:
            data = json.loads(response.stdout)
        except Exception:
            last_error = 'Could not fetch data from Twelve Data'
            continue

        if 'code' in data:
            last_error = data.get('message', 'Twelve Data API Error')
            continue

        if 'values' not in data or not data['values']:
            last_error = 'No values in Twelve Data response'
            continue

        values = [
            {
                'high': entry.get('high'),
                'low': entry.get('low'),
                'close': entry.get('close'),
                'volume': entry.get('volume'),
                'datetime': entry.get('datetime'),
            }
            for entry in data['values']
        ]
        return values, 'twelvedata', None

    return None, None, last_error


def fetch_yfinance(ticker, interval, periods):
    yf_interval = interval_to_yfinance(interval)
    yf_range = yfinance_range_for_interval(interval, periods)

    pycode = r'''
import json
import sys

try:
    import yfinance as yf
except Exception as e:
    print(json.dumps({'error': f'yfinance import failed: {e}'}))
    raise SystemExit(0)

symbol, interval, period = sys.argv[1], sys.argv[2], sys.argv[3]
try:
    try:
        df = yf.Ticker(symbol).history(interval=interval, period=period, auto_adjust=False, prepost=False)
    except Exception as e:
        print(json.dumps({'error': f'yfinance history failed: {e}'}))
        raise SystemExit(0)

    if df is None or df.empty:
        print(json.dumps({'error': 'No values in yfinance response'}))
        raise SystemExit(0)

    rows = []
    for idx, row in df.iterrows():
        high = row.get('High')
        low = row.get('Low')
        close = row.get('Close')
        if high is None or low is None or close is None:
            continue
        volume = row.get('Volume')
        dt_value = None
        try:
            dt_value = idx.isoformat()
        except Exception:
            dt_value = str(idx)
        rows.append({
            'high': float(high),
            'low': float(low),
            'close': float(close),
            'volume': float(volume) if volume is not None else None,
            'datetime': dt_value,
        })

    if not rows:
        print(json.dumps({'error': 'No usable yfinance rows'}))
        raise SystemExit(0)

    rows.reverse()
    print(json.dumps({'values': rows}))
except Exception as e:
    print(json.dumps({'error': str(e)}))
'''

    response = subprocess.run(
        ['python3', '-c', pycode, ticker, yf_interval, yf_range],
        capture_output=True,
        text=True,
    )

    try:
        data = json.loads(response.stdout)
    except Exception:
        return None, None, 'Could not parse yfinance response'

    if data.get('error'):
        return None, None, data['error']

    values = data.get('values', [])
    try:
        periods_int = int(periods)
    except Exception:
        periods_int = 100
    trimmed = values[:periods_int] if periods_int > 0 else values

    if not trimmed:
        return None, None, 'No values in yfinance response'

    return trimmed, 'yfinance', None


def run_breakout(ticker='TSLA', interval='1day', display='1-day', periods='100', output_json='false'):
    values, source, yf_error = fetch_yfinance(ticker, interval, periods)
    td_error = None
    backend_error = None

    if values is None:
        values, source, td_error = fetch_twelvedata(ticker, interval, periods)
        if values is None:
            backend_error = {
                'yfinance': yf_error,
                'twelvedata': td_error,
            }

    if values is None:
        error_payload = {
            'error': 'Unable to fetch market data',
            'details': backend_error,
            'ticker': ticker,
            'display': display,
            'periods': int(periods),
        }
        if output_json == 'true':
            print(json.dumps(error_payload, indent=2))
        else:
            print('Error: Unable to fetch market data')
        return

    output = compute_output(ticker, interval, display, periods, values, source)
    if yf_error and source == 'twelvedata':
        output['fallback_reason'] = yf_error

    if output_json == 'true':
        print(json.dumps(output, indent=2))
        return

    print(f"=== {ticker} Breakout Analysis ({display}, {periods} periods) ===")
    print(f"Last Known Market Price: ${output['current_price']:.2f}")
    print(f"Source: {source}")
    print()

    print('## Upper Resistance Levels')
    for i, r in enumerate(output['upper_resistances'], 1):
        print(f"  {i}. ${r['price']:.2f} (+${r['diff']})")

    print()
    print('## Lower Support Levels')
    for i, r in enumerate(output['lower_supports'], 1):
        print(f"  {i}. ${r['price']:.2f} (-${r['diff']})")

    print()
    print('## Breakout Probability')
    print(f"- UP: {output['probabilities']['up']}%")
    print(f"- DOWN: {output['probabilities']['down']}%")


if __name__ == '__main__':
    ticker = sys.argv[1] if len(sys.argv) > 1 else 'TSLA'
    interval = sys.argv[2] if len(sys.argv) > 2 else '1day'
    display = sys.argv[3] if len(sys.argv) > 3 else '1-day'
    periods = sys.argv[4] if len(sys.argv) > 4 else '100'
    output_json = sys.argv[5] if len(sys.argv) > 5 else 'false'
    run_breakout(ticker, interval, display, periods, output_json)

File: run-brk.c
/* LEGACY / EXPERIMENTAL: The production HackTrader path runs through run-brk.py via run-brk.sh.
 * Do not route live dashboard traffic to this C runner until feature-parity tests exist. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>

typedef struct {
    double price;
    double diff;
} Level;

typedef struct {
    double high;
    double low;
    double close;
    double volume;
    long long timestamp;
    char datetime[32];
} Bar;

typedef struct {
    char td_key[128];
    char massive_key[128];
} Secrets;

double json_extract_double(const char *json, const char *key) {
    char search[64];
    snprintf(search, sizeof(search), "\"%s\":", key);
    char *pos = strstr(json, search);
    if (!pos) return 0.0;
    pos += strlen(search);
    while (*pos == ' ' || *pos == ':') pos++;
    return atof(pos);
}

// Securely read API keys from server-side path
void load_secrets(Secrets *s) {
    memset(s, 0, sizeof(Secrets));
    const char *secure_path = "/var/www/secrets.json";
    const char *local_path = "secrets.json";
    
    FILE *f = fopen(secure_path, "r");
    if (!f) f = fopen(local_path, "r");

    if (!f) {
        strncpy(s->td_key, "1dbc799fe411479686be512ee797a777", sizeof(s->td_key));
        return;
    }

    char line[256];
    while (fgets(line, sizeof(line), f)) {
        char *key_name = NULL;
        if (strstr(line, "\"TWELVEDATA_API_KEY\"")) key_name = s->td_key;
        else if (strstr(line, "\"MASSIVE_API_KEY\"")) key_name = s->massive_key;
        
        if (key_name) {
            char *val_start = strchr(line, ':');
            if (val_start) {
                val_start = strchr(val_start + 1, '\"');
                if (val_start) {
                    val_start++;
                    char *val_end = strchr(val_start, '\"');
                    if (val_end) {
                        size_t len = val_end - val_start;
                        if (len < 128) {
                            strncpy(key_name, val_start, len);
                            key_name[len] = '\0';
                        }
                    }
                }
            }
        }
    }
    fclose(f);
}

// Helper for Massive API interval mapping
void map_massive_interval(const char *interval, int *mul, char *span) {
    if (strcmp(interval, "1min") == 0) { *mul = 1; strcpy(span, "minute"); }
    else if (strcmp(interval, "5min") == 0) { *mul = 5; strcpy(span, "minute"); }
    else if (strcmp(interval, "1h") == 0) { *mul = 1; strcpy(span, "hour"); }
    else { *mul = 1; strcpy(span, "day"); }
}

int fetch_massive(const char *ticker, const char *interval, int periods, const char *api_key, Bar *bars) {
    int mul;
    char span[16];
    map_massive_interval(interval, &mul, span);

    char from_date[11], to_date[11];
    FILE *fp_from = popen("date -u -d '30 days ago' +%Y-%m-%d", "r");
    FILE *fp_to = popen("date -u +%Y-%m-%d", "r");
    if (!fp_from || !fp_to) return 0;
    fgets(from_date, 11, fp_from);
    fgets(to_date, 11, fp_to);
    pclose(fp_from);
    pclose(fp_to);
    
    // Remove newlines
    from_date[strcspn(from_date, "\r\n")] = 0;
    to_date[strcspn(to_date, "\r\n")] = 0;

    char url[512];
    snprintf(url, sizeof(url), "https://api.massive.com/v2/aggs/ticker/%s/range/%d/%s/%s/%s?adjusted=true&sort=desc&limit=50000&apiKey=%s",
             ticker, mul, span, from_date, to_date, api_key);

    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "curl -s '%s'", url);

    FILE *fp = popen(cmd, "r");
    if (!fp) return 0;
    char buffer[1048576]; 
    size_t bytes_read = fread(buffer, 1, sizeof(buffer) - 1, fp);
    buffer[bytes_read] = '\0';
    pclose(fp);

    if (strstr(buffer, "\"status\":\"ERROR\"") || !strstr(buffer, "\"results\":")) {
        return 0;
    }

    int count = 0;
    char *ptr = strstr(buffer, "\"results\":");
    if (!ptr) return 0;
    ptr++;

    while ((ptr = strstr(ptr, "{")) && count < periods) {
        char *end = strstr(ptr, "}");
        if (!end) break;
        char obj[512];
        int len = end - ptr + 1;
        if (len < 512) {
            strncpy(obj, ptr, len);
            obj[len] = '\0';
            // Massive fields: h, l, c, v, t
            bars[count].high = json_extract_double(obj, "h");
            bars[count].low = json_extract_double(obj, "l");
            bars[count].close = json_extract_double(obj, "c");
            bars[count].volume = json_extract_double(obj, "v");
            // Extract timestamp t (it is a long)
            char *t_ptr = strstr(obj, "\"t\":");
            if (t_ptr) {
                t_ptr += 3;
                while (*t_ptr == ' ' || *t_ptr == ':') t_ptr++;
                bars[count].timestamp = atoll(t_ptr);
            }
            count++;
        }
        ptr = end;
    }
    return count;
}

int main(int argc, char *argv[]) {
    if (argc < 5) {
        fprintf(stderr, "Usage: %s <ticker> <interval> <display> <periods> <output_json>\n", argv[0]);
        return 1;
    }

    char *ticker = argv[1];
    char *interval = argv[2];
    char *display = argv[3];
    int periods = atoi(argv[4]);
    int output_json = (argc > 5 && strcmp(argv[5], "true") == 0);

    Secrets s;
    load_secrets(&s);

    Bar *bars = malloc(sizeof(Bar) * periods);
    int count = 0;
    char *source = "unknown";

    // 1. Try Twelve Data
    if (strlen(s.td_key) > 0) {
        char url[512];
        snprintf(url, sizeof(url), "https://api.twelvedata.com/time_series?symbol=%s&interval=%s&apikey=%s&outputsize=%d", 
                 ticker, interval, s.td_key, periods);
        char cmd[1024];
        snprintf(cmd, sizeof(cmd), "curl -s '%s'", url);
        FILE *fp = popen(cmd, "r");
        if (fp) {
            char buffer[1048576]; 
            size_t bytes_read = fread(buffer, 1, sizeof(buffer) - 1, fp);
            buffer[bytes_read] = '\0';
            pclose(fp);
            
            if (!strstr(buffer, "\"code\":") || strstr(buffer, "\"code\":\"200\"")) {
                char *ptr = strstr(buffer, "\"values\":");
                if (ptr) {
                    ptr++;
                    while ((ptr = strstr(ptr, "{")) && count < periods) {
                        char *end = strstr(ptr, "}");
                        if (!end) break;
                        char obj[512];
                        int len = end - ptr + 1;
                        if (len < 512) {
                            strncpy(obj, ptr, len);
                            obj[len] = '\0';
                            bars[count].high = json_extract_double(obj, "high");
                            bars[count].low = json_extract_double(obj, "low");
                            bars[count].close = json_extract_double(obj, "close");
                            bars[count].volume = json_extract_double(obj, "volume");
                            count++;
                        }
                        ptr = end;
                    }
                    if (count > 0) {
                        source = "twelvedata_c_turbo";
                    }
                }
            }
        }
    }

    // 2. Fallback to Massive
    if (count == 0 && strlen(s.massive_key) > 0) {
        count = fetch_massive(ticker, interval, periods, s.massive_key, bars);
        if (count > 0) {
            source = "massive_c_turbo";
        }
    }

    if (count == 0) {
        fprintf(stderr, "No values found from any API source\n");
        free(bars);
        return 1;
    }

    double current_price = bars[0].close;
    double u1 = 1e18, u2 = 1e18;
    for (int i = 0; i < count; i++) {
        if (bars[i].high > current_price) {
            if (bars[i].high < u1) { u2 = u1; u1 = bars[i].high; }
            else if (bars[i].high < u2) { u2 = bars[i].high; }
        }
    }
    double l1 = -1e18, l2 = -1e18;
    for (int i = 0; i < count; i++) {
        if (bars[i].low < current_price) {
            if (bars[i].low > l1) { l2 = l1; l1 = bars[i].low; }
            else if (bars[i].low > l2) { l2 = bars[i].low; }
        }
    }

    double dist_u = (u1 == 1e18) ? 1e18 : u1 - current_price;
    double dist_l = (l1 == -1e18) ? 1e18 : current_price - l1;
    double inv_u = (dist_u > 0) ? 1.0 / dist_u : 0;
    double inv_l = (dist_l > 0) ? 1.0 / dist_l : 0;
    double total_inv = inv_u + inv_l;
    double prob_up = (total_inv > 0) ? (inv_u / total_inv) * 100.0 : 0;
    double prob_down = (total_inv > 0) ? (inv_l / total_inv) * 100.0 : 0;

    double current_vol = bars[0].volume;
    double sum_vol = 0;
    for (int i = 1; i < count; i++) sum_vol += bars[i].volume;
    double avg_vol = (count > 1) ? sum_vol / (count - 1) : 0;
    double vol_ratio = (avg_vol > 0) ? current_vol / avg_vol : 0;

    if (output_json) {
        printf("{\n  \"ticker\": \"%s\",\n  \"current_price\": %.2f,\n", ticker, current_price);
        printf("  \"upper_resistances\": [{\"price\": %.2f, \"diff\": %.2f}, {\"price\": %.2f, \"diff\": %.2f}],\n", u1, u1-current_price, u2, u2-current_price);
        printf("  \"lower_supports\": [{\"price\": %.2f, \"diff\": %.2f}, {\"price\": %.2f, \"diff\": %.2f}],\n", l1, current_price-l1, l2, current_price-l2);
        printf("  \"probabilities\": {\"up\": %.1f, \"down\": %.1f},\n", prob_up, prob_down);
        printf("  \"volume\": {\"current\": %.2f, \"expected\": %.2f, \"ratio\": %.2f},\n", current_vol, avg_vol, vol_ratio);
        
        printf("  \"time_series\": [");
        for (int i = 0; i < count; i++) {
            printf("{\"t\": %lld, \"c\": %.2f}%s", bars[i].timestamp, bars[i].close, (i == count - 1) ? "" : ",");
        }
        printf("],\n");
        
        printf("  \"source\": \"%s\"\n}\n", source);
    } else {
        printf("=== %s Breakout Analysis (%s, %d periods) ===\n", ticker, display, periods);
        printf("Last Known Market Price: $%.2f\n", current_price);
        printf("Source: %s\n\n", source);
        printf("## Upper Resistance Levels\n  1. $%.2f (+%.2f)\n  2. $%.2f (+%.2f)\n\n", u1, u1-current_price, u2, u2-current_price);
        printf("## Lower Support Levels\n  1. $%.2f (-%.2f)\n  2. $%.2f (-%.2f)\n\n", l1, current_price-l1, l2, current_price-l2);
        printf("## Breakout Probability\n- UP: %.1f%%\n- DOWN: %.1f%%\n", prob_up, prob_down);
    }

    free(bars);
    return 0;
}
